### ISSUE TITLE
Accuracy of estimates based on spatial (or in general: structured) REs?

### ISSUE THREAD

**Issue by florianhartig** [glmmTMB/glmmTMB#671]
Hi guys, 

this is not really an issue, more a hint / question: I have supervised a student project on testing an alternative implementation of a spatial CAR Model within sjSDM https://github.com/TheoreticalEcology/s-jSDM

For the project, my student benchmarked the accuracy of the estimated lambda parameter in a standard spatial CAR model with exponential dependency structure against gls, glmmTMB and hand-coded optim implementation. What was noteworthy is that only glmmTMB shows a relatively consistent bias towards smaller lambdas, especially for high true lambdas. I don't want to paste the entire code here, as this was a student project, but here is an example result, points are replicate simulations.

![image](https://user-images.githubusercontent.com/5457753/110128609-48e71600-7dc7-11eb-877f-754dcfd4b6bf.png)

I wonder what you think about this. I haven't checked everything in detail, but I highly suspect that this is replicable. We have also seen some other results (from a different project) that made me suspect that glmmTMB regularizes the REs a bit stronger than other packages. 

Cheers,
Florian

**Comment by mebrooks** [comment]
Thanks for letting us know. This is concerning. Is lambda in this model related to shrinkage of the random effects towards zero? @kaskr @skaug do you have any idea about what might be causing this inaccuracy?

**Comment by florianhartig** [comment]
Hi Mollie, my suspicion was that by shrinking the REs towards zero, lambda inherits some of the shrinkage. I assumed that you don't have any direct shrinkage on lambda, but you have some shrinkage on the REs, right?

**Comment by florianhartig** [comment]
Hi, sorry, yes, lambda is exp(-lambda * distance). OK, will try to create a reproducible example so that you can see if we extracted the parameter correctly.

**Comment by mebrooks** [comment]
Thanks @florianhartig! That would be very helpful.

**Comment by florianhartig** [comment]
Hi guys, OK, what we extract as lambda is exp(-fit.exp$fit$par[4]) in the following example (packages might not all be needed)

```
library(dplyr)
library(ggplot2)
library(Metrics)
library(corpcor)
library(nlme)
library(MVN)
library(DHARMa)
library(useful)
library(glmmTMB)
library(usethis)
library(devtools)
library(knitr)
library(coda)
library(rjags)
library(mvtnorm)
library(rjags)
library(R2jags)
library(mvtnorm)


global.mu = 0

repeat.experiment = data.frame(matrix(NA,500,6))
colnames(repeat.experiment) = c("rep", "Lambda", "Side", "glmm_Time", "glmm_Lambda", "glmm_Intercept")


dist.matrix <- function(side)
{
  row.coords <- rep(1:side, times=side)
  col.coords <- rep(1:side, each=side)
  row.col <- data.frame(row.coords, col.coords)
  D <- dist(row.col, method="euclidean", diag=TRUE, upper=TRUE)
  D <- as.matrix(D)
  return(D)
}


cor.surface <- function(side, global.mu, lambda)
{
  D <- dist.matrix(side)
  # scaling the distance matrix by the exponential decay
  SIGMA <- exp(-lambda*D)
  mu <- rep(global.mu, times=side*side)
  # sampling from the multivariate normal distribution
  M <- matrix(nrow=side, ncol=side)
  M[] <- rmvnorm(1, mu, SIGMA)
  return(M) # list(...)
}

set.seed=226

counter=1
for (g in 1:5 ){    # repeat experiment 5 times
  time_repeat = 
    system.time({
      
      for (h in 1:10 ){     # use 10 different lambdas form 0.2 zil 2.0
        time_glmmTMB = 
          system.time({
            lambda = 0 + (h*0.2)
            l.r <- lambda
            
            for (j in 4:13){   # use different size ( as a side*side 2d array)
              time_side = 
                system.time({
                  side = 2+ 2*j
                  s.r <- side
                  
                  D <- dist.matrix(side)
                  M <- cor.surface(side = side, lambda = lambda, global.mu = global.mu)
                  y <- as.vector(as.matrix(M))
                  
                  
                  new.data1 = 0
                  #n= side*side
                  new.data1        =  data.frame(resp = new.data$y)
                  new.data1$pos    <- numFactor(new.data$col, new.data$row.col)
                  new.data1$group  <- factor(rep(1, new.data$N))
                  new.data1$x      <- new.data$row
                  new.data1$y      <- new.data$col
                  
                  
                  time_glmmTMB = 
                    system.time({
                      fit.exp <- glmmTMB(resp ~ 1 + exp(pos + 0 | group), data = new.data1)
                    })
                  fit.exp$fit
                  
                  repeat.experiment[counter, 1] <- g       # repetitions
                  repeat.experiment[counter, 2] <- lambda  # real lambda
                  repeat.experiment[counter, 3] <- s.r     # side size
                  
                  repeat.experiment[counter, 4] <- time_glmmTMB[3] 
                  repeat.experiment[counter, 5] <- exp(-fit.exp$fit$par[4])
                  repeat.experiment[counter, 6] <- summary(fit.exp)$coefficients$cond[1]
                  
                  counter = counter + 1
                  
                }
                )} 
            
            lambda = 0.2 
            
          }
          )}
      
      
    }
    )}

repeat.experiment

saveRDS(repeat.experiment, file = "data")
```

**Comment by kaskr** [comment]
The model you want to fit has three parameters. The actual fitted model has four parameters because glmmTMB includes an observation level sd (nugget effect) by default.

You can remove it using `disp=~0` although often at the cost of reduced robustness and speed.

**Comment by florianhartig** [comment]
Hi kaskr, just so that I understand: when I type:

`glmmTMB(resp ~ 1 + exp(pos + 0 | group), data = new.data1)`

this actually fits

`glmmTMB(resp ~ 1 + exp(pos + 0 | group) + (1|ID), data = new.data1)`

or is the OLRE also spatially structured?

**Comment by kaskr** [comment]
Almost, but not exactly - I think you wanted to say that

```
glmmTMB(resp ~ 1 + exp(pos + 0 | group), data = new.data1)
```

is equivalent to

```
glmmTMB(resp ~ 1 + exp(pos + 0 | group) + (1|ID), data = new.data1, disp=~0)
```

(to be sure always check the that `logLik()` are the same...)

**Comment by florianhartig** [comment]
I think I'm beginning to understand. So by default, glmmTMB always includes an unstructured residual error, additionally to the structured error, while a standard gls doesn't - is this a fair summary of the situation?

In theory, it makes sense to me to "offer" both a structured and an unstructured error to the data, just to avoid forcing the model to explain certain patterns by a spatial structure. However, it seems it doesn't work very well in practice. I wonder if the reason may be that with equal regularisation on the sd, the unstructured term has at least the potential to produce far more df, so it may possible be easier / less costly under the current likelihood to explain spatial variance by the unstructured term, even if this variance is structured? Of is there any other explanation for the above behavior?

**Comment by kaskr** [comment]
> is this a fair summary of the situation?

Yes

> However, it seems it doesn't work very well in practice.

It should work well as long as the true decorrelation range `~1/lamba` isn't extreme compared to the sampled distances `[min(dist), max(dist)]`. A too short ranged spatial correlation is impossible to distinguish from the nugget. A too long ranged spatially correlated field is impossible to separate from the global mean.

**Comment by bbolker** [comment]
@kaskr, thanks for clearing this up (I should have thought of it: one of my questions was going to be "are you sure the different platforms are using exactly the same statistical model?). @florianhartig , can you ask the student to go back and re-run the simulations with `dispformula = ~0` and confirm that the results fall in line with the other platforms?

**Comment by bbolker** [comment]
I'm closing this for now, feel free to re-open if the simulations do *not* confirm @kaskr's diagnosis.

**Comment by florianhartig** [comment]
Fine for me, although I would note that regardless of the results of setting dispformula = ~0, what we fitted originally is contains the data-generating model as a submodel (nested), so it should have been able to estimate lambda correctly, and probably would for n-> infinity. 

What the results suggest to me is that the current default of having glm( res ~ pred + MVN(cor) + N) is a somewhat dangerous structure for small (= typical) data size n, as the uncorrelated normal may swallow up variation, and thus you get biased estimates for your correlation structure for small n (which is what our simulations show).

**Comment by bbolker** [comment]
Fair enough. It would certainly be possible but IMO not a good idea (at least not without careful consideration) to put in extra logic to make `dispformula=~0` the default for some subset of the structured covariance models, but it could (??) be noted in documentation.

At present we only document covariance structures in `glmmTMB.Rd` and in the vignette. Perhaps we should add a new stand-alone page for covstructs?

I'm going to reopen this and tag it "documentation".

**Comment by florianhartig** [comment]
fyi, with `disp=~0`, we get comparable results to other packages. 

![image](https://user-images.githubusercontent.com/5457753/112440087-2f901480-8d4a-11eb-875c-d26e4334a456.png)
